static/js/public/snap-details/channelMap.js

```diff
@@ -1,294 +1,418 @@
 
-function setTrack(arch, track, packageName, channelMap) {
-  ['stable', 'candidate', 'beta', 'edge'].forEach((risk, i, risks) => {
-    const channelEl = document.getElementById(`js-channel-map-${risk}`);
+class ChannelMap {
+  constructor(selectorString, packageName, channelMapData, defaultTrack) {
+    this.RISK_ORDER = ['stable', 'candidate', 'beta', 'edge'];
+    this.INSTALL_TEMPLATE = document.querySelector('[data-js="install-window"]').innerHTML;
 
-    // channel names in tracks other then latest are prefixed with track name
-    const channelName = track === LATEST ? risk : `${track}/${risk}`;
+    this.packageName = packageName;
+    this.currentTab = 'overview';
 
-    channelEl.querySelector('.js-channel-name').innerHTML = channelName;
-    let channelData = channelMap[channelName];
+    this.openDesktopAttempt = 1;
 
-    // update install instructions
-    let command = `sudo snap install ${packageName}`;
-
-    if (track === LATEST) {
-      if (risk !== 'stable') {
-        command += ` --${risk}`;
-      }
+    if (!defaultTrack) {
+      this.defaultTrack = 'latest';
     } else {
-      command += ` --channel=${track}/${risk}`;
+      this.defaultTrack = defaultTrack;
     }
 
-    channelEl.querySelector('input').value = command;
+    this.selectorString = selectorString;
+    this.channelMapEl = document.querySelector(this.selectorString);
+    this.channelOverlayEl = document.querySelector('.p-channel-map-overlay');
+    this.channelMapData = channelMapData;
 
-    const versionEl = channelEl.querySelector('.p-form-help-text');
-    channelEl.classList.remove('p-channel-map__row--closed');
+    // get architectures from data
+    const architectures = Object.keys(this.channelMapData);
 
-    // show version
-    if (channelData) {
-      versionEl.innerHTML = `Version: ${channelData.version}`;
-    } else {
-      let fallbackRisk;
-      for (let j = 0; j < i; j++) {
-        const channelName = track === LATEST ? risks[j] : `${track}/${risks[j]}`;
-        if (channelMap[channelName]) {
-          fallbackRisk = risks[j];
-        }
-      }
+    // initialize architecture select
+    const archSelect = document.getElementById("js-channel-map-architecture-select");
+
+    archSelect.innerHTML = architectures.map(arch => `<option value="${arch}">${arch}</option>`).join('');
 
-      if (fallbackRisk) {
-        versionEl.innerHTML = `No release in ${risk} channel, using ${fallbackRisk} release.`;
+    archSelect.addEventListener('change', ()=> {
+      this.prepareTable(this.channelMapData[archSelect.value]);
+    });
+
+    this.arch = this.channelMapData['amd64'] ? 'amd64' : architectures[0];
+
+    // To ensure the scope of the event handlers are bound to the class (not the element)
+    // and that 'removeEventListener' works
+    this.closeOnClick = function(event) {
+      this._closeOnClick(event);
+    }.bind(this);
+
+    this.closeOnEscape = function(event) {
```

Comment:
Wouldn't `this._closeOnEscape.bind(this)` do exactly the same?

Reply:
This is removed now for events module